src/frame/compress.rs

```diff
@@ -1,66 +1,361 @@
-use super::MAGIC_NUMBER;
-use super::{BlockSize, Checksum, END_MARK};
-use crate::block::compress::compress as compress_block;
-use std::io::Read;
-use std::io::Write;
-
-/// Configure compression settings.
-pub struct CompressionSettings {
-    #[allow(dead_code)]
-    independent_blocks: bool,
-    /// add block checksum
-    #[allow(dead_code)]
-    block_checksums: Checksum,
-    /// add content checksum
-    #[allow(dead_code)]
-    content_checksum: Checksum,
-    /// sets the block size, see [`BlockSize`]
-    block_size: BlockSize,
+use std::{
+    fmt,
+    hash::Hasher,
+    io::{self, Write},
+};
+use twox_hash::XxHash32;
+
+use crate::block::{
+    compress::compress_internal,
+    hashtable::{HashTable, HashTableU32},
+};
+
+use super::header::{BlockInfo, BlockMode, FrameInfo, BLOCK_INFO_SIZE, MAX_FRAME_INFO_SIZE};
+use super::Error;
+use crate::block::WINDOW_SIZE;
+
+/// A writer for compressing a LZ4 stream.
+///
+/// This `FrameEncoder` wraps any other writer that implements `io::Write`.
+/// Bytes written to this writer are compressed using the [LZ4 frame
+/// format](https://github.com/lz4/lz4/blob/dev/doc/lz4_Frame_format.md).
+///
+/// Writes are buffered automatically, so there's no need to wrap the given
+/// writer in a `std::io::BufWriter`.
+///
+/// To ensure a well formed stream the encoder must be finalized by calling
+/// either `finish` or `try_finish()` methods.
+///
+/// # Example 1
+/// Serializing json values into a compressed file.
+///
+/// ```no_run
+/// let compressed_file = std::fs::File::create("datafile").unwrap();
+/// let mut compressor = lz4_flex::frame::FrameEncoder::new(compressed_file);
+/// serde_json::to_writer(&mut compressor, &serde_json::json!({ "an": "object" })).unwrap();
+/// compressor.finish().unwrap();
+/// ```
+///
+/// # Example 2
+/// Deserializing multiple json values out of a compressed file
```

Comment:
the example code below is serialization. the examples are very similiar, was this intended?

Reply:
Yes they are intentionally similar. I think the second one can be improved to customizer the frame option, or somethinglike that.